lib/sequel/extensions/connection_metrics.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+#
+# The connection_metrics extension enhances a database's
+# connection pool to store metrics about the use of connections.
+# Whenever a connection is acquired or released the number of
+# currently acquired connections is emitted. Same for making new
+# connections and disconnecting them from the DB. Example of use:
+#
+#   DB.extension(:connection_metrics)
+#
+# Note that this extension has been only tested with a
+# connection pool type :threaded. To use it with other
+# connection pool types it might need adjustments.
+#
+# Related module: Sequel::ConnectionMetrics
+
+module Sequel
+  module ConnectionMetrics
+    # Initialize the data structures used by this extension.
+    def self.extended(pool)
+      case pool.pool_type
+      when :single, :sharded_single, :timed_queue, :sharded_threaded, :sharded_timed_queue
+        raise Error.new('cannot load connection_metrics extension if using a connection pool type different than :threaded')
+      end
+
+      pool.instance_exec do
+        sync do
+          @prometheus_updater = CloudController::DependencyLocator.instance.prometheus_updater
+          @connection_info = {}
+        end
+      end
+    end
+
+    private
+
+    def acquire(thread)
+      begin
+        if (conn = super)
+          acquired_at = Time.now
+          @prometheus_updater.increment_gauge_metric(:cc_acquired_db_connections_total, labels: { process_type: })
+          @connection_info[thread] ||= {}
```

Comment:
**question** I don't know ruby concurrency well enough, do we not need any mutex here because the key being read/written to in the hash is unique per thread?

Reply:
As you said, the key is unique for each thread and each thread is only accessing its own key. So I'd argue we do not need a mutex here.